### PR TITLE
Statusline view in search mode or command line mode

### PR DESCRIPTION
**Pull request by tzachar**
When using `cmp-cmdline` in search mode (/ or ?), the custom entries menu is displayed vertically, which obscures the buffer when searching -- especially when using `incsearch`. This new view opens the entries in the statusline, similar to wilder.nvim, which solved the issue.

Some design choices:
1) Enabled by setting `experimental.horizontal_search = true`
2) Only the `abbr` field for each item is used
3) Items are separated by `|`
4) Fuzzy matches etc are highlighted using cmp highligh groups
5) the selected item is highlighted using `CursorLine`, as I don't think statusline supports blending hlgroups.

There might be some bugs, but I have been using it for a while.
Cheers.

### CODE REVIEW

**Comment by tzachar** [comment]
@hrsh7th 

Regarding not using the statusline: is there any benefit to using a float other than blending highlighting groups?

**Comment by max397574** [comment]
@tzachar could you send a screenshot?

**Comment by tzachar** [comment]
@max397574 
here you go
Peek 2022-01-14 13-30

**Comment by tzachar** [comment]
@hrsh7th 
I moved the PR to using a floating window.

**Comment by tzachar** [comment]
@hrsh7th 
Any chance of getting this merged?

**Comment by tzachar** [comment]
@hrsh7th 
I've been using this PR for the past month. It really made a difference for me.
What more is required to get this merged? You mentioned something about configuration change, but did not elaborate.
Thanks.

**Comment by Shougo** [comment]
Hm...  I think the feature should be configurable.
And the title should be `Statusline view in search mode or command line mode`.

**Comment by hrsh7th** [comment]
I'll check it. Sorry for the late.

**Comment by hrsh7th** [comment]
My random thoughts are here. I want to discuss these points.

##### The configuration option shouldn't be experimental

I think if we ship this feature, we don't plan to remove this. So it isn't experimental.
But I can't think reasonable option for this. Please let me your ideas.

##### It should be worked in insert-mode and normal cmdline-mode.

I don't think it's useful that it's worked in insert-mode. But it's consistent.

##### UI should be the same as native wildmenu.

But it's depending on to the personal preferences...

```lua
cmp.setup {
  ...
}

cmp.setup.cmdline('/', {
  ...
  some_reasonabole_option = true
  ...
})
```

**Comment by hrsh7th** [comment]
BTW, It's useful :)

File: lua/cmp/view.lua [hunk 1]
```diff
@@ -189,10 +192,14 @@ view._get_entries_view = function(self)
...
-    self.custom_entries_view.event:on('change', function()
+    local v = self.custom_entries_view
+    if vim.tbl_contains({ '/', '?' }, vim.fn.getcmdtype()) and c.experimental.horizontal_search then
+      v = self.statusline_entries_view
+    end
+    v.event:on('change', function()
       self:on_entry_change()
     end)
-    return self.custom_entries_view
+    return v
```
Comment (hrsh7th) [inline lua/cmp/view.lua:202]:
```lua
cmp.setup {
  appearance = {
    entries_view = 'custom' | 'statusline' | 'native', -- the `native` will be removed in the future.
  }
}
```

I'm afraid the `appearance` and `entries_view` are good wording or not. (I'm always afraid of such wording decisions...

**Comment by hrsh7th** [comment]
BTW, It this behavior intented? It isn't the same as native wildmenu but not bad. It's interesting.

https://user-images.githubusercontent.com/629908/151798718-64c9b904-1201-4b16-aa30-a37dbbc3fc2b.mp4

File: lua/cmp/view.lua (hunk 1 above)
Comment (gegoune) [reply lua/cmp/view.lua:202]: How about `display.layout` or `window.layout`?

**Comment by tzachar** [comment]
which behaviour do you refer to?

**Comment by tzachar** [comment]
> My random thoughts are here. I want to discuss these points.
> 
> ##### The configuration option shouldn't be experimental
> I think if we ship this feature, we don't plan to remove this. So it isn't experimental. But I can't think reasonable option for this. Please let me your ideas.
> 
> ##### It should be worked in insert-mode and normal cmdline-mode.
> I don't think it's useful that it's worked in insert-mode. But it's consistent.

If its not useful then why support it?

> ##### UI should be the same as native wildmenu.
> But it's depending on to the personal preferences...

Can you be more specific? 

> ```lua
> cmp.setup {
>   ...
> }
> 
> cmp.setup.cmdline('/', {
>   ...
>   some_reasonabole_option = true
>   ...
> })
> ```

File: lua/cmp/view.lua (hunk 1 above)
Comment (hrsh7th) [reply lua/cmp/view.lua:202]: We already have `cmp.setup.cmdline`.
You can set appearance option for only cmdline-mode.

Comment (tzachar) [reply lua/cmp/view.lua:202]: Ok. So maybe we we can add the appearance config option both in general and in cmdline mode.
Makes sense?

Comment (hrsh7th) [reply lua/cmp/view.lua:202]: @tzachar Yes. I think this feature is needless in insert-mode. But it's useful to keep the consistency.

What do you think about this?
```
cmp.setup {
  view = {
    entries = 'custom' | 'statusline' | 'native',
  }
}
```

Comment (hrsh7th) [reply lua/cmp/view.lua:202]: Hm... `display` / `appearance` / `layout` / `window` / `view` / `ui`...

**Comment by hrsh7th** [comment]
>> UI should be the same as native wildmenu.
>> But it's depending on to the personal preferences...

> Can you be more specific?

The following screencast uses `native wildmenu`.

The native wildmenu is

- Does not contains `|` chars.
- The items are separated by two spaces.
- The default highlight is `StatusLine` (maybe... I can't find actual one from vim help)
- The active item highlight is `WildMenu`.

https://user-images.githubusercontent.com/629908/151822440-03ace504-9d2c-4c3a-92fd-22cee5ad88af.mp4

**Comment by hrsh7th** [comment]
Hm... But I think we shouldn't use wildmenu highlights. We should use pum highlights.

File: lua/cmp/view/statusline_entries_view.lua [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  self:init()
+
+  self.entries_win = window.new()
+  self.entries_win:option('conceallevel', 2)
+  self.entries_win:option('concealcursor', 'n')
+  self.entries_win:option('cursorlineopt', 'line')
+  self.entries_win:option('foldenable', false)
+  self.entries_win:option('wrap', false)
+  self.entries_win:option('scrolloff', 0)
+  self.entries_win:option('winhighlight', 'Normal:Pmenu,FloatBorder:Pmenu,CursorLine:PmenuSel,Search:None')
```
Comment (hrsh7th) [inline lua/cmp/view/statusline_entries_view.lua:42]: We should create new highlight groups but it's ok at the moment.

File: lua/cmp/view/statusline_entries_view.lua [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+end
+
+statusline_entries_view.abort = function(self)
+  feedkeys.call('', 'n', function()
+    self:close()
+  end)
+end
+
+statusline_entries_view.draw = function(self)
+  local entries_buf = self.entries_win:get_buffer()
```
Comment (hrsh7th) [inline lua/cmp/view/statusline_entries_view.lua:163]: Probably, I'll try to refactor it using the `&sidescrolloff` and `vim.api.nvim_win_set_cursor` for managing viewport after merge this PR.

File: lua/cmp/view.lua [hunk 4]
```diff
@@ -4,6 +4,7 @@ local event = require('cmp.utils.event')
 local keymap = require('cmp.utils.keymap')
 local docs_view = require('cmp.view.docs_view')
 local custom_entries_view = require('cmp.view.custom_entries_view')
+local statusline_entries_view = require('cmp.view.statusline_entries_view')
```
Comment (hrsh7th) [inline lua/cmp/view.lua:7]: I think the filename and varname shouldbe `wildmenu_entries_view`. (What do you think about it?)

File: lua/cmp/view.lua (hunk 1 above)
Comment (hrsh7th) [reply lua/cmp/view.lua:202]: I think we can merge this after solve this.

File: lua/cmp/view.lua (hunk 4 above)
Comment (tzachar) [reply lua/cmp/view.lua:7]: I dont really have an opinion here. I dont mind changing if you do.

File: lua/cmp/view.lua (hunk 1 above)
Comment (hrsh7th) [reply lua/cmp/view.lua:202]: OK. I decided.
I'll organize the configuration options after merging this PR.

**Comment by hrsh7th** [comment]
I will merge it once and organize it. Is that okay?

**Comment by tzachar** [comment]
I think its ready to be commited now.
Config is like you suggested:
```lua
  view = {                  
    entries = 'custom'      
  }                         
````

and in cmdline:
```lua
cmp.setup.cmdline('/', {         
  view = {                       
    entries = 'wildmenu',        
    separator = '|',             
  },                            
}
```

**Comment by hrsh7th** [comment]
I think the `separator` is needed only for `entries` and the value of `wildmenu`. So it's should be contained to the `{ name = 'wildmenu', separator = '|' }`

```lua
cmp.setup {
  view = {
    entries = { name = 'wildmenu', separator = '|' } -- the user can also specify the `wildmenu` literal string.
  }
}
```

**Comment by tzachar** [comment]
> I think the `separator` is needed only for `entries` and the value of `wildmenu`. So it's should be contained to the `{ name = 'wildmenu', separator = '|' }`
> 
> ```lua
> cmp.setup {
>   view = {
>     entries = { name = 'wildmenu', separator = '|' } -- the user can also specify the `wildmenu` literal string.
>   }
> }
> ```
done

**Comment by hrsh7th** [comment]
LGTM.

**Comment by tzachar** [comment]
@hrsh7th 
On last thing I forgot: what about adding some documentation, pointing users to the new feature?

**Comment by dsully** [comment]
Does this change address the issue of the "normal" vertical menu crossing over the status line?

**Comment by tzachar** [comment]
probably. It overlays the status line with the entries.

**Comment by dsully** [comment]
Hm. Sounds like a different issue then. I want the ("normal") vertical completion menu, but to have it render above the status line.

**Pull request merged**